crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -149,6 +152,81 @@ impl Solution {
         self.user_trades().next().is_none()
     }
 
+    pub fn merge(&self, other: &Self) -> Result<Self, MergeError> {
+        // We can only merge solutions from the same solver
+        if self.solver.account().address() != other.solver.account().address() {
+            return Err(MergeError::Incompatible("Solvers"));
+        }
+
+        // Solutions should not settle the same order twice
+        let uids = self
+            .trades()
+            .iter()
+            .filter_map(|trade| match trade {
+                Trade::Fulfillment(fulfillment) => Some(fulfillment.order().uid),
+                Trade::Jit(_) => None,
+            })
+            .collect::<HashSet<_>>();
+        if other
+            .trades()
+            .iter()
+            .filter_map(|trade| match trade {
+                Trade::Fulfillment(fulfillment) => Some(fulfillment.order().uid),
+                Trade::Jit(_) => None,
+            })
+            .any(|other_uid| uids.contains(&other_uid))
+        {
+            return Err(MergeError::DuplicateTrade);
+        }
+
+        // Solution prices need to congruent, i.e. there needs to be a unique factor to
+        // scale all common tokens from one solution into the other.
+        let scaling_factors = scaling_factors(&self.prices, &other.prices);
+        let factor = match scaling_factors.len() {
```

Comment:
This logic seems pretty awkward. I think calling the function `common_scaling_factor() -> Result<BigRational>` or sth like that and moving that check into the function would be easier to understand.

I would also move `if factor < 1` there.

Reply:
Wasn't able to get the if `factor < 1` logic abstracted but I think this is much simpler to read now.